### PR TITLE
[python] Protect against huge enum-of-strings input

### PR DESCRIPTION
**Pull request by johnkerl**
**Issue and/or context:** #3353 [[sc-59407]](https://app.shortcut.com/tiledb-inc/story/59407)

**Changes:**

As proposed on #3353 [[sc-59407]](https://app.shortcut.com/tiledb-inc/story/59407)

**Notes for Reviewer:**

### CODE REVIEW

File: apis/python/src/tiledbsoma/io/ingest.py [hunk 1]
```diff
@@ -1068,6 +1070,130 @@ def _create_or_open_coll(
...
+        column_names=["soma_joinid"]
+    ).concat()
+    previous_join_ids = set(
+        int(e)
+        for e in get_dataframe_values(previous_sjids_table.to_pandas(), SOMA_JOINID)
+    )
+    mask = [e.as_py() not in previous_join_ids for e in arrow_table[SOMA_JOINID]]
+    arrow_table = arrow_table.filter(mask)
+
+    # This is a redundant, failsafe check. The appebd-mode registrar already
```
Comment (nguyenv) [inline apis/python/src/tiledbsoma/io/ingest.py:1126]:
```suggestion
    # This is a redundant, failsafe check. The append-mode registrar already
```

File: apis/python/src/tiledbsoma/io/conversions.py [hunk 2]
```diff
@@ -56,27 +65,17 @@ def to_tiledb_supported_array_type(name: str, x: _MT) -> _MT:
...
+    # If the column is categorical-of-string of high cardinality, we declare
+    # this is likely a mistake, and it will definitely lead to performance
+    # issues in subsequent processing.
+    if isinstance(x, pd.Series) and isinstance(x.dtype, pd.CategoricalDtype):
+        # Heuristic number
+        if (
+            pandas.api.types.is_string_dtype(x)
+            and len(x.cat.categories) > STRING_DECAT_THRESHOLD
+        ):
+            return x.astype(str)
```
[the remark this answers is not in the document]
Comment (johnkerl) [reply apis/python/src/tiledbsoma/io/conversions.py:77]: What do you think @bkmartinjr ?

File: apis/python/src/tiledbsoma/io/ingest.py [hunk 3]
```diff
@@ -1212,6 +1331,7 @@ def _write_dataframe_impl(
             # Nominally, nil id_column_name only happens for uns append and we do not append uns,
             # which is a concern for our caller. This is a second-level check.
             raise ValueError("internal coding error: id_column_name unspecified")
+        # XXX
```
[the remark this answers is not in the document]
Comment (johnkerl) [reply apis/python/src/tiledbsoma/io/ingest.py:1334]: It's not needed at all.

Sorry about having included that in the review! :(

File: apis/python/src/tiledbsoma/io/conversions.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply apis/python/src/tiledbsoma/io/conversions.py:77]: I would not - I would instead just add it to the help/docstrings that this is the default behavior.

I don't find these warnings particularly useful, as most of this code runs in production pipelines. Better to document the behavior in the API docs, IMHO.

File: apis/python/src/tiledbsoma/io/conversions.py [hunk 4]
```diff
@@ -47,36 +60,26 @@ def _to_tiledb_supported_dtype(dtype: _DT) -> _DT:
...
```
Comment (bkmartinjr) [inline apis/python/src/tiledbsoma/io/conversions.py:81]: this is a bug - assumes all categoricals are of type `str`.  They can be any primitive type, e.g., `int`, `float`, `bool, etc.

Comment (johnkerl) [reply apis/python/src/tiledbsoma/io/conversions.py:81]: Thanks @bkmartinjr -- I'll make a follow-on PR -- this one is "Protect against huge enum-of-strings input" -- I'll generalize the follow-on to "Protect against huge enum-of-anything input"

**Pull request merged**